doit/cmd_base.py

```diff
@@ -310,30 +310,64 @@ def _load_from(cmd, namespace, cmd_list):
         return task_list, doit_config
 
 
-class ModuleTaskLoader(TaskLoader):
+class TaskLoader2(TaskLoader):
+    """Interface of task loaders with new-style API.
+
+    The default implementation assumes tasks are loaded from a namespace, mapping identifiers to
+    elements like functions (like task generators) or constants (like configuration values).
+
+    This API update separates the loading of the configuration and the loading of the actual tasks,
+    which enables additional elements to be available during task creation.
+    """
+    def __init__(self):
+        super().__init__()
+        self.namespace = None
+
+    def setup(self, opt_values):
+        """Delayed initialization.
+
+        To be implemented if the data is needed by derived classes.
+
+        :param opt_values: (dict) with values for cmd_options
+        """
+
+    def load_doit_config(self):
+        """Load doit configuration."""
+        return loader.load_doit_config(self.namespace)
+
+    def load_tasks(self, cmd, pos_args):
+        """Load tasks.
+
+        :return: (List[Task])
+        :param cmd: (doit.cmd_base.Command) current command being executed
+        :param pos_args: (list str) positional arguments from command line
+        """
+        return loader.load_tasks(self.namespace, self.cmd_names, cmd.execute_tasks)
+
+
+class ModuleTaskLoader(TaskLoader2):
     """load tasks from a module/dictionary containing task generators
     Usage: `ModuleTaskLoader(my_module)` or `ModuleTaskLoader(globals())`
     """
-    cmd_options = ()
-
     def __init__(self, mod_dict):
         super(ModuleTaskLoader, self).__init__()
-        self.mod_dict = mod_dict
-
-    def load_tasks(self, cmd, params, args):
-        return self._load_from(cmd, self.mod_dict, self.cmd_names)
+        if inspect.ismodule(mod_dict):
+            self.namespace = dict(inspect.getmembers(mod_dict))
+        else:
+            self.namespace = mod_dict
 
 
-class DodoTaskLoader(TaskLoader):
+class DodoTaskLoader(TaskLoader2):
     """default task-loader create tasks from a dodo.py file"""
     cmd_options = (opt_dodo, opt_cwd, opt_seek_file)
 
-    def load_tasks(self, cmd, params, args):
-        dodo_module = loader.get_module(
-            params['dodoFile'],
-            params['cwdPath'],
-            params['seek_file'])
-        return self._load_from(cmd, dodo_module, self.cmd_names)
+    def setup(self, opt_values):
+        # lazily load namespace from dodo file per config parameters:
+        self.namespace = loader.get_module(
+            opt_values['dodoFile'],
+            opt_values['cwdPath'],
+            opt_values['seek_file'],
+        ).__dict__
```

Comment:
Original code is using ``inspect.getmembers``. Any subtle difference? I guess better not rely on ``__dict__``.